test/counter_server_example/src/counter_server_example.erl

```diff
@@ -0,0 +1,261 @@
+%%% @doc
+%%%
+%%% Example for using vnet to simulate a cluster of nodes, one of
+%%% which serves as a counter server.
+%%%
+%%% All _test functions demonstrate scenarios of using the server. See
+%%% their documentation for more info.
+%%%
+%%% All _no_vnet_test functions do not use vnet and their purpose is
+%%% to check the core implementation of the counter server.
+%%%
+%%% @end
+%%%-------------------------------------------------------------------
+
+-module(counter_server_example).
+
+%%%-------------------------------------------------------------------
+
+-include_lib("eunit/include/eunit.hrl").
+
+-concuerror_options(
+   [ {instant_delivery, false}
+   , assertions_only
+   , {ignore_error, deadlock}
+   ]).
+
+%%%-------------------------------------------------------------------
+
+-export(
+   [ setup_server/0
+   , once_valid_client/0
+   , twice_valid_client/0
+   , invalid_client/0
+   , twice_valid_same_gen_client/0
```

Comment:
For me it's very hard to follow the purpose of the functions in this module. If I understand correctly, you have:
* EUnit test functions (`*_test/0`) that will get auto-exported by EUnit, and test the no-vnet scenarios.
* Concuerror test functions exported here that test the vnet-scenarios.
* Functions that exported so that they can be used together with `vnet:rpc/4`.
* Other helper functions.

I would prefer if these classes of functions would be coupled together in the file with appropriate headers and preferably with some common naming convention to make them easy to identify.

Reply:
The module is written as an 'essay'. The _test functions are all Concuerror tests and I am indeed taking advantage of EUnit to have them exported.

There is a comment at the top of the file making an attempt to explain that.